### DISCUSSION TITLE
What is the logic for incremental generation of index files

### DISCUSSION THREAD

**Discussion by qintian-qe** [oracle/opengrok#4262] [General]
When incrementally generating an index for multiple git warehouse codes under repo management, whether to first perform a global traversal to determine whether the code has changed, or directly determine whether the code is updated based on the git log, and then perform an index update？

**Comment by qintian-qe** [comment]
When incrementally generating an index for multiple git warehouse codes under repo management, whether to first perform a global traversal to determine whether the code has changed, or directly determine whether the code is updated based on the git log, and then perform an index update？

**Reply by vladak** [reply]
The following is assuming that projects are enabled: when the index is created for a project for the first time, the directory of the related directory is recursively traversed and the list of files to process is collected. On subsequent reindex, it depends whether history cache was enabled. If yes, and all the repositories for that project support that kind of operation (e.g. they are all Git), the list of files is extracted using the SCM method to avoid traversing the whole directory tree. If any of these conditions is not met, the indexer will fall back to the directory traversal method.

Why do you ask ?

**Reply by qintian-qe** [reply]
> The following is assuming that projects are enabled: when the index is created for a project for the first time, the directory of the related directory is recursively traversed and the list of files to process is collected. On subsequent reindex, it depends whether history cache was enabled. If yes, and all the repositories for that project support that kind of operation (e.g. they are all Git), the list of files is extracted using the SCM method to avoid traversing the whole directory tree. If any of these conditions is not met, the indexer will fall back to the directory traversal method.
> 
> Why do you ask ?

Previously, when creating a new task, I found that creating a historycache was slow, so I didn't create a historycache file, causing me to traverse the directory tree again every time I made a code update

**Reply by vladak** [reply]
It's a trade-off. I guess this can be made to work even without history cache, however currently the repository detection and inner workings in the indexer are tied to the history cache being on.
